pkg/cloudprovider/providers/aws/aws.go

```diff
@@ -1436,51 +1446,92 @@ func (c *Cloud) InstanceType(ctx context.Context, nodeName types.NodeName) (stri
 }
 
 // getCandidateZonesForDynamicVolume retrieves  a list of all the zones in which nodes are running
-// It currently involves querying all instances
 func (c *Cloud) getCandidateZonesForDynamicVolume() (sets.String, error) {
 	// We don't currently cache this; it is currently used only in volume
 	// creation which is expected to be a comparatively rare occurrence.
 
 	// TODO: Caching / expose v1.Nodes to the cloud provider?
 	// TODO: We could also query for subnets, I think
+	zones, err := c.getCachedCandidateZonesForDynamicVolume()
+	if err != nil {
+		glog.V(2).Infof("getCachedCandidateZonesForDynamicVolume() failed, falling back to legacy method: %v", err)
 
-	filters := []*ec2.Filter{newEc2Filter("instance-state-name", "running")}
+		filters := []*ec2.Filter{newEc2Filter("instance-state-name", "running")}
 
-	instances, err := c.describeInstances(filters)
+		instances, err := c.describeInstances(filters)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(instances) == 0 {
+			return nil, fmt.Errorf("no instances returned")
+		}
+
+		zones := sets.NewString()
+
+		for _, instance := range instances {
```

Comment:
The filtering logic could be shared for both methods (but this would required generating a new object for each instance/node based on where it comes from: informer or ec2 describe API call)
It would be easier to understand I think

Reply:
Again, this is the legacy path, we only take it if the Node-Informer is not ready. I would favour removing the whole block if we can. Discussion for upstream.

I leave the legacy path untouched as a safe fallback. We can be more adventurous upstream.